### PR TITLE
Customize cell content, fixes #12

### LINKED ISSUE #12: Cell (get/formatter/renderCell)
**Enhancement** 

In dgrid 1, the `get` function was used to return a text value and `formatter`was used to return HTML - inserted in the standard cell wrapper. In addition to this, renderCell could be used to return your own DOM which would use the value of `get` but `formatter` wouldn't be called as part of this lifecycle.

`get` is a way to prevent HTML injection that may come from user data and `formatter` is a way to mark the HTML as potentially safe.

Ken, in his brain dump, noted that this decision for this API was mostly a legacy holdover . He proposed re-implementing these methods as renderCell, renderText, and renderHtml.

`DNode` already combines renderText and renderHtml (through the `innerHTML` property). Also, because dgrid 2 uses a registry, `renderCell` is accomplished instead by registering a different widget for 'cell'.

Our implementation should look like this:

* If `render` is used, it is the only method that is called with item and column data and no value is found
* If `get` is used, it is called with item and column data to find the value
* Otherwise, the value is not used, a property lookup is performed on the item data
* If `renderValue` is used, it is called with the value and the item and column data
* The content created by `render`, `renderValue`, or a stringified value is passed to the cell so that the cell has both the value and the content

### PR DESCRIPTION
**Pull request by pottedmeat**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [ ] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

**Description:**

Adds new callbacks to the `Column` interface, updates `Row` to use the new `Column` definition to find `Cell` content, and adds unit tests to check the new behavior.

### CODE REVIEW

**Review by dylans — COMMENTED** [review]
For the things I review for typically, this looks fine, though someone else should review and approve before it lands. Thanks!

**Review by kitsonk — APPROVED** [review]

File: src/interfaces.d.ts
Comment (msssk) [inline src/interfaces.d.ts]: I still don't see the need for `id` - `field` is enough.

File: src/interfaces.d.ts
Comment (msssk) [inline src/interfaces.d.ts:78]: Am I understanding correctly that dgrid requires data objects to look like this:
```javascript
{
    id: <idValue>,
    data: {
        <... data properties>
    }
}
```

Are we not able to continue dgrid 1's design of accepting flatter objects (with support for `idProperty` as well):

```javascript
{
    id: <idValue>,
    <... data properties>
}
```

File: src/Cell.ts
Comment (msssk) [inline src/Cell.ts:11]: What is the purpose of `HasValue` for `Cell`? It's not used in the rendering... is it envisioned that one might want to read the value of `value` from a `Cell` instance?

**Review by msssk — APPROVED** [review]

File: src/Cell.ts
Comment (pottedmeat) [reply src/Cell.ts:11]: That's my thought, yes.

File: src/interfaces.d.ts
Comment (pottedmeat) [reply src/interfaces.d.ts:78]: Right now, the object structure is simple, but `ItemProperties` adds much more moving forward.

Comment (msssk) [reply src/interfaces.d.ts:78]: I think it would be very good if dgrid 2 is able to continue supporting data in the same format as dgrid 1 unless there's a compelling reason not to - is there?

While Dojo's data solutions have evolved over the years, the basic format of a data object has been consistent for years. What is the compelling need to change it?

Also - considering dgrid 2 requires a data provider, are my concerns relevant? Can one store be used for consumers that like the old data format, and also for dgrid, with the differences implemented in the data provider?

Comment (pottedmeat) [reply src/interfaces.d.ts:78]: This doesn't really have anything to do with the data - it's what the data provider gives to the grid. `ArrayDataProvider` still supports simple data with `idProperty`.

Comment (msssk) [reply src/interfaces.d.ts:78]: ok great!

File: src/interfaces.d.ts
Comment (agubler) [inline src/interfaces.d.ts]: We should have some documentation for exported interfaces, will help explain the intent of the properties and functions.

File: src/Row.ts
Comment (agubler) [inline src/Row.ts]: I'm not sure I get why we need `render` and `renderValue`, couldn't there just be a `render` (perhaps a better name, to not conflate with a widgets `render` function) and just have an optional third parameter `value?: V`?

File: src/Row.ts
Comment (agubler) [inline src/Row.ts:42]: Perhaps would be nicer it could be simplified a bit (does assume consolidating the two render functions):

```ts
let value = item.data[ field || id] ? String(item.data[ field || id ]) : '';
if (column.get) {
    value = column.get(item, column);
}

let content: DNode = value;
if (column.render) {
    content = column.render(item, column, value);
}
```

**Review by agubler — COMMENTED** [review]
A few initial comments

File: src/Row.ts
Comment (pottedmeat) [reply src/Row.ts]: I've been going back and forth on this, and it's been discussed in the referenced issue. My intent with render was to have a function that indicates it will not be consuming any data - adding an action button, for example. Absent this, we will still attempt to read a value from the underlying data - which isn't horrible but creates a side-effect that may be unexpected and could (though it would be unlikely) create problems.

Some alternatives I'd been considering:

* Add something like an enum that can be assigned to `field` that would indicate no value should be read from the data object
* Allow `get` to be either a function or a value - perhaps rename this property to `value`

 What do you think of these alternatives?

Comment (agubler) [reply src/Row.ts]: I don't fully understand the problem with having a single function but with an API that passes the value field, this leaves it up to the consumer to decide how to implement the customisation and choose to use the value or not.

Comment (pottedmeat) [reply src/Row.ts]: The most general use-case in our experience is having a custom `get` and a custom `renderValue`. In this situation, the user will probably only care about the value and not the item or column, so having value first would be ideal. If they don't care about having a value, having value last would be ideal. I'm messing around with this signature and I'm not 100% happy with having value last.

Comment (agubler) [reply src/Row.ts]: Then do you think it would be more ergonomic to have an options style API? 
```ts
interface CustomRenderOptions {
    item: any;
    column: Column;
    value: any;
}

// consumer implements function and only uses what they need from the options
function customRender(options: CustomRenderOptions): DNode {
    // do something with either options.item, options.column or options.value
    // or any combination.
}
```

Comment (agubler) [reply src/Row.ts]: I see you have updated the API, to normalise the `render` functions, I would still consider a different name than `render` (to remove potential confusion)

Comment (agubler) [reply src/Row.ts]: I am still not sure about needing `get` to be a function when the existing `render` function can be used for that.

Comment (pottedmeat) [reply src/Row.ts]: I had considered the options API, but with this covering the most common use cases in the first arguments, it doesn't seem worth the overhead.

Do you have any suggestions for a better name?

`render` may be used without `get` and `get` may be used without `render`. Combining everything into a single function will tightly couple the customization our users have traditionally done as well as force users to implement what are pretty standard defaults.

Comment (agubler) [reply src/Row.ts]: With the `options` I'd just expect consumers to destructure what is needed at the top of their custom renderer function (which I think is ergonomic).

```ts
const { value } = options;
```

Both `get` and `render` return a `DNode` and are used to effectively populate `content` (i'm not sure why cell needs `value` now there is `content`) that is used by `Cell`? Just feels like having two functions that end up as the same thing - `content` is more confusing than flexible (would be for me), because I can return a string either based on item and column.

So just having a API for `customCellRenderer(item: any, column: Column): DNode {}`, should cover all the scenarios needed?

```ts
const { customCellRenderer, field = column.id } = column;
let content = item.data[ field ] ? String(item.data[ field ]) : '';
if (customCellRenderer) {
    content = customCellRenderer(item, column);
}
```

Means a consumer can return a 

1) Static string:

```ts
function customCellRenderer(item: any, column: Column): DNode {
     return 'static-string';
}
```

2) Dynamic string based on the item

```ts
function customCellRenderer(item: any, column: Column): DNode {
    const { field = column.id } = column;
    const value = item[field];
    return `Miss ${value}`;
}
```

3) A custom DNode

```ts
function customCellRenderer(item: any, column: Column): DNode {
    return v('button', { /* some properties */});
}
```

Comment (agubler) [reply src/Row.ts]: Unless I've missed a scenario that having the two separate functions covers.

Comment (pottedmeat) [reply src/Row.ts]: `get` is specific to finding the value to be used in a cell (the combination of the item and column) and is meant to present an alternative to the default method of retrieving this value from the item based on a field.

`render` is meant to take this value and turn it into the UI that will be used inside the cell. `render` is reusable, whereas `get` is not. So `render` could appear across all columns, while `get` is unique to each column.

I made it so that `get` could return a `DNode` so that if someone really wanted to do everything in a single function, they could (though I agree it isn't great practice, it just seemed the best way to do things if we wanted to avoid having separate render methods for when you do and do not care about the cell's value).

I hope this is enough to explain why `get` and `render`, `value` and `content` are two separate ideas that seem to encourage separate levels of customization. If that can be agreed upon, I can remove the `DNode` return from `get` so it retains this clear separation and I guess I can also make the `render` function accept an object, though I wish there was a way to do this without so much overhead.

File: src/interfaces.d.ts
Comment (kitsonk) [inline src/interfaces.d.ts]: Alphabetise please... :-D

Comment (pottedmeat) [reply src/interfaces.d.ts]: I think I'm reading the style guide on interface property ordering correctly - listing properties then methods. I couldn't find anything on ordering documentation for interfaces but is that what you're referring to here that needs alphebetized?

Comment (kitsonk) [reply src/interfaces.d.ts]: yes, we need to make it explicit, we have been trying to alphabetize things by default, as when they get added to, it becomes difficult to find stuff in big interefaces... properties then methods, yup, but alphabetical within those groups

Comment (pottedmeat) [reply src/interfaces.d.ts]: Is it not alphabetized within each group right now?

Comment (kitsonk) [reply src/interfaces.d.ts]: `get` is listed as a property with a value of method (or a string), where `render` is an optional method... not the same class of properties

**Review by kitsonk — APPROVED** [review]

**Pull request merged**
